timelock-server/src/test/java/com/palantir/atlasdb/timelock/lock/ExclusiveLockTests.java

```
@@ -0,0 +1,177 @@
+
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+
+import org.junit.Test;
+
+public class ExclusiveLockTests {
+
+    private static final UUID REQUEST_1 = UUID.randomUUID();
+    private static final UUID REQUEST_2 = UUID.randomUUID();
+    private static final UUID REQUEST_3 = UUID.randomUUID();
+
+    private final ExclusiveLock lock = new ExclusiveLock();
+
+    @Test
+    public void canLockAndUnlock() {
+        lockSynchronously(REQUEST_1);
+        lock.unlock(REQUEST_1);
+    }
+
+    @Test
+    public void lockIsExclusive() {
+        lockSynchronously(REQUEST_1);
+
+        CompletableFuture<Void> future = lockAsync(REQUEST_2);
+        assertNotComplete(future);
+    }
+
+    @Test
+    public void lockCanBeObtainedAfterBeingUnlocked() {
+        lockSynchronously(REQUEST_1);
+        lock.unlock(REQUEST_1);
+
+        lockSynchronously(REQUEST_2);
+    }
+
+    @Test
+    public void queuedRequestObtainsLockAfterBeingUnlocked() {
+        lockSynchronously(REQUEST_1);
+        CompletableFuture<Void> future = lockAsync(REQUEST_2);
+
+        unlock(REQUEST_1);
+
+        assertCompleteSuccessfully(future);
+    }
+
+    @Test
+    public void multipleQueuedRequestsCanObtainLock() {
+        lockSynchronously(REQUEST_1);
+        CompletableFuture<Void> future2 = lockAsync(REQUEST_2);
+        CompletableFuture<Void> future3 = lockAsync(REQUEST_3);
+
+        unlock(REQUEST_1);
+
+        assertCompleteSuccessfully(future2);
+        assertNotComplete(future3);
+
+        unlock(REQUEST_2);
+
+        assertCompleteSuccessfully(future3);
+    }
+
+    @Test
+    public void unlockByNonHolderThrows() {
+        lockSynchronously(REQUEST_1);
+
+        assertThatThrownBy(() -> unlock(UUID.randomUUID()))
+                .isInstanceOf(IllegalStateException.class);
+        assertThat(lock.getCurrentHolder()).isEqualTo(REQUEST_1);
+    }
+
+    @Test
+    public void unlockByWaiterThrows() {
```

Hmm not sure we actually care about that case. It's actually a product bug if unlock is ever called by anyone but the holder (which is why we throw `IllegalStateException` rather than returning false or something)